src/compiler/checker.ts

```diff
@@ -10550,20 +10550,21 @@ namespace ts {
         // constituents, in which case the isPartial flag is set when the containing type is union type. We need
         // these partial properties when identifying discriminant properties, but otherwise they are filtered out
         // and do not appear to be present in the union type.
-        function getUnionOrIntersectionProperty(type: UnionOrIntersectionType, name: __String): Symbol | undefined {
+        function getUnionOrIntersectionProperty(type: UnionOrIntersectionType, name: __String, skipObjectFunctionPropertyAugment?: boolean): Symbol | undefined {
             const properties = type.propertyCache || (type.propertyCache = createSymbolTable());
             let property = properties.get(name);
             if (!property) {
-                property = createUnionOrIntersectionProperty(type, name);
-                if (property) {
+                property = createUnionOrIntersectionProperty(type, name, skipObjectFunctionPropertyAugment);
+                // Dont set the result since it might not be correct if skipping lookup from property of Object and function type
+                if (property && !skipObjectFunctionPropertyAugment) {

```

Comment:
Shouldn't we have a secondary cache here for these props? There's a possibility we do quite a bit of work during the global setup phase.